Updated NTT to be similar to the rewritten FFT

Chillee opened the pull request:
Since we rewrote the FFT and got a massive speedup, seems strange that the NTT has such a different implementation from the FFT.

This PR updates the NTT to be written in the style of the new FFT, and gets a 7-10x boost in performance from doing so: https://ideone.com/QJwGfE

This also includes an update to the fuzz test for ntt - A. because the new FFT implementation returns empty vector if either one of the input polynomials is empty, and that's the API I copied for the NTT, and B. because the new NTT now requires that all inputs be within the mod range.

This is still somewhat WIP - I have a couple questions/issues I need to investigate more (and some notes)
1. The conditional modular adds/subtracts add about 30% in performance UNLESS `O3` is enabled. If `O3` is enabled then it slows it down by about 50%. In this case, `O3` is harming the performance of the conditional modular adds/subtracts. It goes from about 200 ms to 450 ms with `O3`. I don't understand why, and I can't replicate it in an isolated example.

2. Is it possible to shorten this:
```cpp
    int curL = (mod - 1) >> 2;
    for (int k = 2; k < n; k *= 2) {
        int z[] = {1, modpow(root, curL)};
        curL >>= 1;
        rep(i, k, 2 * k) rt[i] = mul(rt[i / 2], z[i & 1]);
    }
```
3. Handling the output vector. It's currently handled in a redundant way, and I feel like it can be cleaned up nicely.

4. Re: the modular operations, it may be a good idea to inline them into the code (especially add/subtract since they're only used once).

I also need to test it more thoroughly, I'm not absolutely confident there's no issues with certain sizes and what not.

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -18,37 +18,41 @@ const ll mod = (119 << 23) + 1, root = 3; // = 998244353
+    copy(all(a), begin(inl)), copy(all(b), begin(inr));
+    ntt(inl, rt, rev, n), ntt(inr, rt, rev, n);
+    int invN = modpow(n, mod - 2);
+    rep(i, 0, n) out[i] = mul(mul(inl[i], inr[i]), invN);
```
`out[-i & (n-1)] = ...` to avoid the reverse

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -18,37 +18,41 @@ const ll mod = (119 << 23) + 1, root = 3; // = 998244353
+    reverse(out.begin() + 1, out.begin() + n);
+    ntt(out, rt, rev, n);
+    rep(i, 0, sz(res)) res[i] = out[i];
+    return out;
```
`return res;`, or `return {out.begin(), out.end() + s};` for some `s` declared higher up (the FFT needs to convert complex->real, but here things are simpler)

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
+    vi out(n), rt(n, 1);
+    vi rev(n);
+    rep(i, 0, n) rev[i] = (rev[i / 2] | (i & 1) << L) / 2;
+    int curL = (mod - 1) >> 2;
```
`mod / 4`

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
+    rep(i, 0, n) rev[i] = (rev[i / 2] | (i & 1) << L) / 2;
+    int curL = (mod - 1) >> 2;
+    for (int k = 2; k < n; k *= 2) {
+        int z[] = {1, modpow(root, curL)};
```
`curL /= 2` as argument (and use `/ 2` above)?

or `curL /= 2` below; it shaves off one character and compiles the same

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
+        curL >>= 1;
+        rep(i, k, 2 * k) rt[i] = mul(rt[i / 2], z[i & 1]);
+    }
+    copy(all(a), begin(inl)), copy(all(b), begin(inr));
```
an alternative to this would be
```
vi inl(all(a)), inr(all(b));
inl.resize(n); inr.resize(n);
```
(it adds some memory allocations but hopefully it's not that much slower). OTOH that might not be a LOC win when putting all `vi` declarations on the same line.

Chillee replied:
You can also just do `vi inl(a), inr(b)`, right? so it's also a character win, at least.

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
+    int L = 32 - __builtin_clz(sz(res)), n = 1 << L;
+    vi inl(n), inr(n);
+    vi out(n), rt(n, 1);
+    vi rev(n);
```
same line for all of these

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
+            rep(j, 0, k) {
+                int z = mul(rt[j + k], a[i + j + k]);
+                a[i + j + k] = sub(a[i + j], z);
+                a[i + j] = add(a[i + j], z);
```
Did this improve performance much over the previous version? This requires inputs to be within [0, mod), and adds add/sub functions.

Chillee replied:
https://ideone.com/emwmZA

About 25% gain. We can inline the add/sub functions.

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
-	vl roots(sz(x), 1), temp = roots;
-	rep(i,1,sz(x)) roots[i] = roots[i-1] * e % mod;
-	ntt(&x[0], &temp[0], &roots[0], sz(x), 1);
+int mul(int a, int b) { return (ll)a * b % mod; }
```
You get a -Wconversion warning here, and in a bunch of other places. -Wconversion is used by KACTL's default compile macro, and super important for not getting overflows in number theory-heavy code, so please make sure the code compiles without such warnings. Possibly by using `ll` for all modded numbers (`int` and `ll` should be about as fast on 64-bit systems, and even on 32-bit ones I think perf might be dominated by modulo operations? So with some luck this won't slow anything down. But please do benchmark...)

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
+    copy(all(a), begin(inl)), copy(all(b), begin(inr));
+    ntt(inl, rt, rev, n), ntt(inr, rt, rev, n);
+    int invN = modpow(n, mod - 2);
+    rep(i, 0, n) out[i] = mul(mul(inl[i], inr[i]), invN);
```
can we reuse `inl` instead of having a new `out` vector?

Chillee replied:
I don't easily see how if we're doing the `[-i & (n-i)]` for the bit reversal - perhaps you know how though.

simonlindholm replied:
oh, good point, pick your poison :)

simonlindholm commented on fuzz-tests/ntt.cpp:
```diff
@@ -21,49 +21,52 @@ ll modpow(ll a, ll e) {
 }
 
 typedef vector<ll> vl;
```
Ooops, I started commenting on the fuzz test rather than the original code. Is it possible to use an `#include` here? Perhaps with some trickery for getting the right `mod`, e.g. `namespace ignored { #include "modpow.h" } #include "ntt.h"` or `#define`s

simonlindholm commented:
> This PR updates the NTT to be written in the style of the new FFT, and gets a 7-10x boost in performance from doing so: https://ideone.com/QJwGfE

That's very nice! I thought mod operations would dominate perf and make this a smaller win, but apparently not. Is that number for 64-bit?

> The conditional modular adds/subtracts add about 30% in performance UNLESS O3 is enabled. If O3 is enabled then it slows it down by about 50%. In this case, O3 is harming the performance of the conditional modular adds/subtracts. It goes from about 200 ms to 450 ms with O3. I don't understand why, and I can't replicate it in an isolated example.

-O3.  ¯\\\_(ツ)\_/¯ I wouldn't worry so much about it, but checking the assembly might be interesting (`gcc -S`).

> Re: the modular operations, it may be a good idea to inline them into the code (especially add/subtract since they're only used once).

Yes please!

Chillee commented:
Responded to all comments.

Chillee commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -18,41 +18,33 @@ const ll mod = (119 << 23) + 1, root = 3; // = 998244353
-        rep(i, k, 2 * k) rt[i] = mul(rt[i / 2], z[i & 1]);
+        ll z[] = {1, modpow(root, curL/=2)};
+        rep(i, k, 2 * k) rt[i] = rt[i / 2]* z[i & 1]%mod;
     }
```
The roots are calculated in this way in FFT not only for numerical precision, but also for cache efficiency.

Chillee commented:
I think all review comments have been addressed on this.

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -18,41 +18,33 @@ const ll mod = (119 << 23) + 1, root = 3; // = 998244353
-                a[i + j] = add(a[i + j], z);
+                ll z = rt[j + k] * a[i + j + k] % mod;
+                a[i + j + k] = z > a[i + j] ? a[i + j] - z + mod : a[i + j] - z;
+                a[i + j] = a[i + j] + z >= mod ? a[i + j] + z - mod : a[i + j] + z;
```
I don't see this resolved -- we should definitely do it since this is over the character limit. Did you forget to push some commit?

Chillee replied:
Whups... :sweat_smile:

simonlindholm commented:
See individual comment. Another thing to do before merging is to convert to tabs rather than spaces -- with tabs = 2 spaces in the pdf, 4-space indentation would be inconsistent. (Everyone misses this, so maybe we should just convert the whole KACTL repo over to spaces... but that's for another PR.)

Chillee commented:
Whups, you're right - didn't push. My bad.

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -12,43 +12,39 @@
-	ntt(c, true); c.resize(s); return c;
+	ntt(inl, rt, rev, n), ntt(inr, rt, rev, n);
+	ll invN = modpow(n, mod - 2);
+	rep(i,0,n) out[-i&(n-1)] = (inl[i] * inr[i] % mod)* invN%mod;
```
no need for the parentheses. Space before/after %; if that causes the line to exceed 63 chars, you can rename invN to inv

Speaking of renames, we could also rename inl/inr to e.g. L/R, which is harder to typo

Chillee replied:
I don't like capital variable names :sweat_smile: although your rationale makes sense.

I changed the spacing and renamed `invN` to `inv`. If you think it's worth it to change it to `L/R`, can you do it?

simonlindholm replied:
Not seeing this change. I can make the L/R change; I think it's worth it. (I somewhat like capital letter names, actually - it makes them stand out, and they aren't that much harder to type. I almost always use them for problem inputs.)

Chillee replied:
I also use them for problem inputs, but *only* problem inputs. So it feels weird to me to use them elsewhere. Amended the commit with the actual change...

Chillee replied:
Also made the `L`/`R` change since you're set on it.

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -12,43 +12,39 @@
-vl conv(vl a, vl b) {
-	int s = sz(a) + sz(b) - 1; if (s <= 0) return {};
-	int L = s > 1 ? 32 - __builtin_clz(s - 1) : 0, n = 1 << L;
-	if (s <= 200) { // (factor 10 optimization for |a|,|b| = 10)
```
Have you tested whether there exists a cut-off for the new implementation as well? probably isn't necessary any more, but interesting to know.

Chillee replied:
Brute force now starts to surpass FFT at N=32.

For N<=4 it's 10x  faster :^)
For N=8 it's 5x faster
For N=16 it's 4x faster
For N=32 it's 1.6x faster
For N=64 it's 1.5x slower.

Definitely not worth it I think.

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -12,43 +12,39 @@
+			rep(j,0,k) {
+				ll z = rt[j + k] * a[i + j + k] % mod, &ai = a[i + j];
+				a[i + j + k] = z > ai ? ai - z + mod : ai - z;
+				ai = ai + z >= mod ? ai + z - mod : ai + z;
```
`ai += ai + z >= mod ? z - mod : z;`

simonlindholm commented on content/numerical/NumberTheoreticTransform.h:
```diff
@@ -12,43 +12,39 @@
  */
```
The doc comment still says `May return negative values.`; remove that and mention that inputs have to be in [0, mod) instead.

Update author/date as well.

Chillee commented:
uhh.... That was not on purpose.

Chillee commented:
Ok, I believe I fixed the git disaster - I also fixed a naming issue caused by naming the vector `L` (I renamed the other variable B for "bits", or something like that).

Chillee commented:
One more note here, I think we might want to add something to make sure `NTT` isn't compiled with `-O3`. Not sure whether we should add something like this: https://stackoverflow.com/questions/31373885/how-to-change-optimization-level-of-one-function or just keep it as documentation.
